app/org/sagebionetworks/bridge/services/ConsentServiceImpl.java

```diff
@@ -42,189 +46,191 @@
+            e.printStackTrace(); // REMOVEME
+            int len = account.getConsentSignatureHistory(subpopGuid).size();
+            account.getConsentSignatureHistory(subpopGuid).remove(len-1);
             accountDao.updateAccount(study, account);
             throw e;
         }
-        
         // Save supplemental records, fire events, etc.
         if (userConsent != null){
             activityEventService.publishEnrollmentEvent(user.getHealthCode(), userConsent);
         }
         optionsService.setEnum(study, user.getHealthCode(), SHARING_SCOPE, sharingScope);
+        
+        updateSessionConsentStatuses(user, subpopGuid, true);
+        studyEnrollmentService.incrementStudyEnrollment(study, user);
+        
         if (sendEmail) {
-            MimeTypeEmailProvider consentEmail = new ConsentEmailProvider(study, user, 
+            MimeTypeEmailProvider consentEmail = new ConsentEmailProvider(study, subpopGuid, user, 
                 consentSignature, sharingScope, studyConsentService, consentTemplate);
 
             sendMailService.sendEmail(consentEmail);
         }
-        user.setConsent(true);
         return user;
     }
 
     @Override
-    public boolean hasUserConsentedToResearch(StudyIdentifier studyIdentifier, User user) {
-        checkNotNull(user, Validate.CANNOT_BE_NULL, "user");
-        checkNotNull(studyIdentifier, Validate.CANNOT_BE_NULL, "studyIdentifier");
-
-        return userConsentDao.hasConsented(user.getHealthCode(), studyIdentifier);
-    }
-
-    @Override
-    public boolean hasUserSignedActiveConsent(StudyIdentifier studyIdentifier, User user) {
-        checkNotNull(user, Validate.CANNOT_BE_NULL, "user");
-        checkNotNull(studyIdentifier, Validate.CANNOT_BE_NULL, "studyIdentifier");
-
-        UserConsent userConsent = userConsentDao.getActiveUserConsent(user.getHealthCode(), studyIdentifier);
-        StudyConsentView mostRecentConsent = studyConsentService.getActiveConsent(studyIdentifier);
-
-        if (mostRecentConsent != null && userConsent != null) {
-            return userConsent.getConsentCreatedOn() == mostRecentConsent.getCreatedOn();
-        } else {
-            return false;
-        }
+    public List<ConsentStatus> getConsentStatuses(ScheduleContext context) {
+        checkNotNull(context);
+        checkNotNull(context.getHealthCode());
+        
+        return subpopService.getSubpopulationForUser(context).stream().map(subpop -> {
+            boolean consented = userConsentDao.hasConsented(context.getHealthCode(), subpop.getGuid());
+            boolean mostRecent = hasUserSignedActiveConsent(context.getHealthCode(), subpop.getGuid());
+            return new ConsentStatus.Builder().withName(subpop.getName()).withGuid(subpop.getGuid())
+                    .withRequired(subpop.isRequired()).withConsented(consented).withSignedMostRecentConsent(mostRecent)
+                    .build();
+        }).collect(BridgeCollectors.toImmutableList());
```

Comment:
It would be trivially easy to convert this code into building a map instead of a list. That way, it's maps all the way down, and we don't have to do awkward conversions between maps and lists, and we don't have to use that awkward forSubpopulation() method.